src/common/ad_tx.c

```diff
@@ -0,0 +1,552 @@
+	}
+	blob_decref(tx->tx_blob);
+	return rc;
+}
+
+/**
+ * snapshot data from address to either redo or undo log
+ */
+int
+ad_tx_snap(struct ad_tx *tx, void *addr, daos_size_t size, uint32_t flags)
+{
+	bool	 undo = (flags & AD_TX_UNDO);
+	bool	 redo = (flags & AD_TX_REDO);
+
+	if (redo == undo)
+		return -DER_INVAL;
+
+	if (addr == NULL || size == 0 || size > UMEM_ACT_PAYLOAD_MAX_LEN)
+		return -DER_INVAL;
+
+	if (tx == NULL) /* noop */
+		return 0;
+
+	if (undo) {
+		struct umem_act_item	*it_undo;
+
+		D_ALLOC_ACT(it_undo, UMEM_ACT_COPY, size);
+		if (it_undo == NULL)
+			return -DER_NOMEM;
+
+		memcpy(it_undo->it_act.ac_copy.payload, addr, size);
+		it_undo->it_act.ac_copy.addr = blob_ptr2addr(tx->tx_blob, addr);
+		it_undo->it_act.ac_copy.size = size;
+		AD_TX_ACT_ADD(tx, it_undo, ACT_UNDO);
+	} else {
+		struct umem_act_item	*it_redo = NULL;
+
+		D_ALLOC_ACT(it_redo, UMEM_ACT_COPY, size);
+		if (it_redo == NULL)
+			return -DER_NOMEM;
+
+		memcpy(it_redo->it_act.ac_copy.payload, addr, size);
+		it_redo->it_act.ac_copy.addr = blob_ptr2addr(tx->tx_blob, addr);
+		it_redo->it_act.ac_copy.size = size;
+		AD_TX_ACT_ADD(tx, it_redo, ACT_REDO);
+	}
+	return 0;
+}
+
+/**
+ * copy data from buffer @ptr to storage address @addr, both old and new data can be saved
+ * for TX redo and undo.
+ */
+int
+ad_tx_copy(struct ad_tx *tx, void *addr, daos_size_t size, void *ptr, uint32_t flags)
+{
+	struct umem_act_item	*it_undo = NULL, *it_redo = NULL;
+	bool			 ptr_only = flags & AD_TX_COPY_PTR;
+	bool			 undo = flags & AD_TX_UNDO;
+	bool			 redo = flags & AD_TX_REDO;
+
+	if (addr == NULL || ptr == NULL || size == 0 || size > UMEM_ACT_PAYLOAD_MAX_LEN)
+		return -DER_INVAL;
+
+	if (tx == NULL) {
+		memcpy(addr, ptr, size);
+		return 0;
+	}
+
+	if (undo) {
+		D_ALLOC_ACT(it_undo, UMEM_ACT_COPY, size);
+		if (it_undo == NULL)
+			return -DER_NOMEM;
+	}
+
+	if (redo) {
+		if (ptr_only)
+			D_ALLOC_ACT(it_redo, UMEM_ACT_COPY_PTR, size);
+		else
+			D_ALLOC_ACT(it_redo, UMEM_ACT_COPY, size);
+		if (it_redo == NULL && it_undo != NULL) {
+			D_FREE(it_undo);
+			return -DER_NOMEM;
+		}
+	}
+
+	if (undo) {
+		memcpy(it_undo->it_act.ac_copy.payload, addr, size);
+		it_undo->it_act.ac_copy.addr = blob_ptr2addr(tx->tx_blob, addr);
+		it_undo->it_act.ac_copy.size = size;
+		AD_TX_ACT_ADD(tx, it_undo, ACT_UNDO);
+	}
+
+	if (redo) {
+		if (ptr_only) {
+			it_redo->it_act.ac_copy_ptr.addr = blob_ptr2addr(tx->tx_blob, addr);
+			it_redo->it_act.ac_copy_ptr.ptr = (uintptr_t)ptr;
+			it_redo->it_act.ac_set.size = size;
+		} else {
+			memcpy(it_redo->it_act.ac_copy.payload, ptr, size);
+			it_redo->it_act.ac_copy.addr = blob_ptr2addr(tx->tx_blob, addr);
+			it_redo->it_act.ac_copy.size = size;
+		}
+		AD_TX_ACT_ADD(tx, it_redo, ACT_REDO);
+	}
+
+	return 0;
+}
+
+static void
+assign_integer(void *addr, int size, uint32_t val)
+{
+	switch (size) {
+	default:
+		D_ASSERT(0);
+	case 1:
+		*((uint8_t *)addr) = (uint8_t)val;
+	case 2:
+		*((uint16_t *)addr) = (uint16_t)val;
+	case 4:
```

Comment:
(style)  Possible switch case/default not preceeded by break or fallthrough comment